### ISSUE TITLE
Missing gems using --standalone mode when gem is specified in gemspec and Gemfile.

### ISSUE THREAD

**Issue by JonRowe** [rubygems/rubygems#4754]
### Describe the problem as clearly as you can

When running `bundle install --standalone` on `2.2.23` gems which are specified in both a `<some-gem>.gemspec` (albeit as development dependencies) and the `Gemfile` are not installed, and thus cannot be bin stubbed or required.

### Post steps to reproduce the problem

- Install bundler 2.2.23 with rubygems 3.2.23
- Add a development dependency to a gemspec e.g.

```
Gem::Specification.new do |spec|
  spec.name = "bundler-issue"
  spec.version = "0.0.1"
  spec.authors = ["Jon Rowe"]
  spec.homepage = "https://github.com/JonRowe/bundler-issue"
  spec.license = "MIT"
  spec.summary = "An example repo"

  spec.files = `git ls-files -- lib/*`.split("\n")
  spec.files += %w[README.md]
  spec.test_files = []
  spec.rdoc_options = ["--charset=UTF-8"]
  spec.require_paths = ["lib"]

  spec.required_ruby_version = '>= 1.8.7'

  spec.add_development_dependency "rake", "> 10.0.0"
end
```

- Use the gemspec in the Gemfile and also specify a version of the same gem

```
source 'https://rubygems.org'

gemspec

gem 'rake', '>= 12.3.3'
```

- Run `bundle install --standalone`
- Observe the lack of rake.
- Run `bundle binstubs rake`
- Observe the failure

Alternatively, here is a repo I used to isolate this issue: https://github.com/JonRowe/bundler-issue
Complete with a failing action run: https://github.com/JonRowe/bundler-issue/actions/runs/1025955181

### Which command did you run?

`bundle install --standalone`

### What were you expecting to happen?

All gems to be installed

### What actually happened?

```
Run bundle install --standalone
Resolving dependencies...
Using bundler 2.2.23
Using bundler-issue 0.0.1 from source at `.`
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Bundled gems are installed into `./bundle`
```

## Environment

```
Bundler             2.2.23
  Platforms         ruby, x86_64-darwin-19
Ruby                3.0.2p107 (2021-07-07 revision 0db68f023372b634603c74fca94588b457be084c) [x86_64-darwin19]
  Full Path         /Users/jon/.rvm/rubies/ruby-3.0.2/bin/ruby
  Config Dir        /Users/jon/.rvm/rubies/ruby-3.0.2/etc
RubyGems            3.2.22
  Gem Home          /Users/jon/.rvm/gems/ruby-3.0.2@Scratch
  Gem Path          /Users/jon/.rvm/gems/ruby-3.0.2@Scratch:/Users/jon/.rvm/rubies/ruby-3.0.2/lib/ruby/gems/3.0.0
  User Home         /Users/jon
  User Path         /Users/jon/.gem/ruby/3.0.0
  Bin Dir           /Users/jon/.rvm/gems/ruby-3.0.2@Scratch/bin
Tools
  Git               2.31.1
  RVM               1.29.12 (latest)
  rbenv             not installed
  chruby            not installed
  rubygems-bundler  (1.4.5)
```

## Bundler Build Metadata

```
Built At          2021-07-09
Git SHA           e863a3905d
Released Version  true
```

## Bundler settings

```
jobs
  Set for the current user (/Users/jon/.bundle/config): 4
path
  Set for your local app (/Users/jon/Code/Scratch/bundler-issue/.bundle/config): "bundle"
```

### Gemfile

```ruby
source 'https://rubygems.org'

gemspec

gem 'rake', '>= 12.3.3'
```

### Gemfile.lock

```
PATH
  remote: .
  specs:
    bundler-issue (0.0.1)

GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  x86_64-darwin-19

DEPENDENCIES
  bundler-issue!

BUNDLED WITH
   2.2.23
```

### bundler-issue.gemspec

```ruby
# coding: utf-8
lib = File.expand_path('../lib', __FILE__)
$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)

Gem::Specification.new do |spec|
  spec.name = "bundler-issue"
  spec.version = "0.0.1"
  spec.authors = ["Jon Rowe"]
  spec.homepage = "https://github.com/JonRowe/bundler-issue"
  spec.license = "MIT"
  spec.summary = "An example repo"

  spec.files = `git ls-files -- lib/*`.split("\n")
  spec.files += %w[README.md]
  spec.test_files = []
  spec.rdoc_options = ["--charset=UTF-8"]
  spec.require_paths = ["lib"]

  spec.required_ruby_version = '>= 1.8.7'

  spec.add_development_dependency "rake", "> 10.0.0"
end
```

**Comment by JonRowe** [comment]
This originated from RSpec's CI runs, which have `rake` listed as a development dependency in the gemspec but is overridden in the Gemfile to allow various versions for differing Rubies.

**Comment by deivid-rodriguez** [comment]
Sounds like #4748, which will be fixed by #4751. Sorry for the regression! :pray:

**Comment by JonRowe** [comment]
Oh it does, sorry for the noise, there is a slight difference in that there were never any warnings the way we used it that I could see.
